src/rule_definition_tools.jl

```diff
@@ -340,13 +357,13 @@ function tuple_expression(primal_sig_parts)
     end
 end
 
-function _nondiff_rrule_expr(primal_sig_parts, primal_invoke)
+function _nondiff_rrule_expr(primal_sig_parts, primal_invoke; isfunctor=false)
     tup_expr = tuple_expression(primal_sig_parts)
     primal_name = first(primal_invoke.args)
     pullback_expr = Expr(
         :function,
         Expr(:call, propagator_name(primal_name, :pullback), :_),
-        Expr(:tuple, NO_FIELDS, Expr(:(...), tup_expr))
+        Expr(:tuple, isfunctor ? DoesNotExist() : NO_FIELDS, Expr(:(...), tup_expr))
```

Comment:
Ratherr than passing in `isfunctor` to do this in the macro defintition
could we do something in the body that compiles away like check `T<:Function && fieldcount(T) == 0`?

Reply:
I don't think that check would work well for types, since e.g. `DataType`s do have fields, but we don't care about the fields for the purposes of Zygote.
Of course, we could also just always return `DoesNotExist()` here, so `@non_differentiable` would just always imply that the function argument has no well-defined derivative, where it's not really important whether it actually has fields or not.

Reply:
Yeah, we could just return `DoesNotExist` for all.
Things without fields are not pertubable so DoesNotExist is not invalid for them.
I think i would be down with that change.

Reply:
We do technically document that the derivative for the primal itself is `NO_FIELDS`, but I agree with your assessment that it's unlikely to break any actual code. Would you say that's still fine for a patch release?

Reply:
@oxinabox patch is fine right?

Reply:
Bump. Since this is holding up JuliaDiff/ChainRules.jl#358

Reply:
I am pretty sure it is fine